src/coreclr/tools/superpmi/superpmi/streamingsuperpmi.cpp

```diff
@@ -0,0 +1,225 @@
+    {
+        LogError("Failed to open file '%s'. GetLastError()=%u", o.streamFile, GetLastError());
+        return 1;
+    }
+
+    // Just one worker for now... all method selection done via stream file
+    //
+    o.workerCount = 1;
+    o.indexes = nullptr;
+    o.indexCount = -1;
+    o.hash = nullptr;
+    o.offset = -1;
+    o.increment = -1;
+
+    // The method context reader handles skipping any unrequested method contexts
+    // Used in conjunction with an MCI file, it does a lot less work...
+    MethodContextReader* reader =
+        new MethodContextReader(o.nameOfInputMethodContextFile, o.indexes, o.indexCount, o.hash, o.offset, o.increment);
+    if (!reader->isValid())
+    {
+        return (int)SpmiResult::GeneralFailure;
+    }
+
+    JitInstance* jit = nullptr;
+
+    char line[2048];
+    const char* const seps = "!";
+    char *next = nullptr;
+
+    // Syntax is dddd { ! <jit-option>=value }*
+    // Likes starting with '#' are ignored
+    //
+    while (fgets(line, sizeof(line), streamFile) != nullptr)
+    {
+        for (int i = 0; i < sizeof(line); i++)
+        {
+            if (line[i] == '\n' || line[i] == '\r')
+            {
+                line[i]= 0;
+                break;
+            }
+        }
+
+        size_t len = strlen(line);
+
+        LogDebug("[streaming] Request: '%s'", line);
+
+        if (line[0] == '#')
+        {
+            continue;
+        }
+
+        if (strncmp(line, "quit", 4) == 0)
+        {
+            LogDebug("[streaming] Quitting");
+            break;
+        }
+
+        char* tok = strtok_r(line, seps, &next);
+        const int index = atoi(tok);
+
+        if (index == 0)
+        {
+            LogDebug("[streaming] Stopping");
+            break;
+        }
+
+        LogDebug("[streaming] Method %d", index);
+
+        LightWeightMap<DWORD,DWORD>* baseForceJitOptions = o.forceJitOptions;
+        LightWeightMap<DWORD,DWORD>* forceJitOptions = nullptr;
+        bool skip = false;
+
+        while ((tok = strtok_r(nullptr, seps, &next)))
+        {
+            if (forceJitOptions == nullptr)
+            {
+                if (baseForceJitOptions == nullptr)
+                {
+                    forceJitOptions = new LightWeightMap<DWORD, DWORD>();
+                }
+                else
+                {
+                    forceJitOptions = new LightWeightMap<DWORD, DWORD>(*baseForceJitOptions);
+                }
+            }
+
+            bool added = AddJitOption(forceJitOptions, tok);
+
+            if (!added)
+            {
+                LogInfo("[streaming] unable to parse option '%s'", tok);
+                skip = true;
+                break;
+            }
+        }
+
+        if (skip)
+        {
+            continue;
+        }
+
+        LogDebug("[streaming] Launching...");
+
+        // launch this instance...
+        //
+        reader->Reset(&index, 1);
+        MethodContextBuffer mcb = reader->GetNextMethodContext();
```

Comment:
In your scenario, it looks like you re-run compilations on the same MC many times. Instead of re-reading (and re-hydrating) the same MC over and over, why don't you first see if the `mc` you have is the one you want, and re-use it. You could imagine having a small cache of loaded `mc` if that makes sense. You still need to delete (and null out) `mc->cr` after every run.

Reply:
Seems like it's the `MethodContextBuffer` that is the key thing to cache.

Let me try this.

I was also going to have the client run MCS and produce a smaller MCH with just the methods it wants to repeatedly evaluate, but am not sure how to match up the indexes from the big and small sets. Does MCH subsetting keep things in relative order or in index relative order or is there no relationship?

That is, do we end up with
```
;; old->new index
3,5,2 -> 1,2,3 (arg order)
3,5,2 -> 2,3,1 (index order)
```
or something other?

Reply:
Not sure about the order; you'll have to check.

As for caching, doesn't the MethodContext (not MethodContextBuffer) contain the hydrated types?

Reply:
Yes, added caching.

The only messy part was that MC's always need to have some kind of CR attached, so I now delete the old one and attach a new empty one.